source/Core/Src/gui.cpp

```diff
@@ -937,7 +972,35 @@ static void displayMenu(size_t index) {
   // Draw symbol
   // 16 pixel wide image
   // 2 pixel wide scrolling indicator
-  OLED::drawArea(96 - 16 - 2, 0, 16, 16, (&SettingsMenuIcons[(16 * 2) * index]));
+  static TickType_t menuSwitchLoopTick = 0;
+  static size_t     menuCurrentIndex   = sizeof(rootSettingsMenu) + 1;
+  if (!animOpenState) {
+    if (menuCurrentIndex != index) {
+      menuCurrentIndex   = index;
+      menuSwitchLoopTick = xTaskGetTickCount();
+    }
+    if (xTaskGetTickCount() - menuSwitchLoopTick < systemSettings.animationSpeed) {
+    loop:
+      OLED::drawFilledRect(OLED_WIDTH - 16 - 2, 0, OLED_WIDTH - 2, OLED_HEIGHT, true);
+      OLED::drawArea(OLED_WIDTH - 16 - 2, 0, 16, OLED_HEIGHT, (&SettingsMenuIcons[index][(16 * 2) * 0]));
+    } else if (xTaskGetTickCount() - menuSwitchLoopTick < systemSettings.animationSpeed * 2) {
+      OLED::drawFilledRect(OLED_WIDTH - 16 - 2, 0, OLED_WIDTH - 2, OLED_HEIGHT, true);
+      OLED::drawArea(OLED_WIDTH - 16 - 2, 0, 16, OLED_HEIGHT, (&SettingsMenuIcons[index][(16 * 2) * 1]));
+    } else {
+      if (systemSettings.animationLoop)
+        if (xTaskGetTickCount() - menuSwitchLoopTick < systemSettings.animationSpeed * 3) {
+          OLED::drawFilledRect(OLED_WIDTH - 16 - 2, 0, OLED_WIDTH - 2, OLED_HEIGHT, true);
+          OLED::drawArea(OLED_WIDTH - 16 - 2, 0, 16, OLED_HEIGHT, (&SettingsMenuIcons[index][(16 * 2) * 2]));
+        } else {
+          menuSwitchLoopTick = xTaskGetTickCount();
+          goto loop;
+        }
+      else {
+        OLED::drawFilledRect(OLED_WIDTH - 16 - 2, 0, OLED_WIDTH - 2, OLED_HEIGHT, true);
+        OLED::drawArea(OLED_WIDTH - 16 - 2, 0, 16, OLED_HEIGHT, (&SettingsMenuIcons[index][(16 * 2) * 2]));
+      }
+    }
+  }
```

Comment:
I would love to see not requiring the goto :)
I think we can simplify a lot of this logic by using the modulo operator to get the index

So doing `((xTaskGetTickCount-start)/duration)%3)`
And then if the `xTaskGetTickCount-start > systemSettings.animationSpeed * 3` explicitly draw the last frame instead unless loop turned on

But otherwise logic is sound

Reply:
IDK how to write like that, please give me example, or completed code
sry

Reply:
Anyway, I think that you can't do without goto, I tried without it, and the picture disappeared in the animation for a couple of milliseconds

Reply:
@Ralim check code now

Reply:
There are still too many if conditionals than necessary in my opinion. If you use modulo like @Ralim suggested you should be able to simplify half of them, and the logic would also be more straightforward.

Reply:
Ok, now I know how to use

Reply:
That looks a _lot_ better; thank you :)